trace-forward/src/Trace/Forward/Network/Acceptor.hs

```diff
@@ -0,0 +1,157 @@
+import           Ouroboros.Network.IOManager (withIOManager)
+import           Ouroboros.Network.Snocket (Snocket, localAddressFromPath, localSnocket, socketSnocket)
+import           Ouroboros.Network.Socket (AcceptedConnectionsLimit (..),
+                                           SomeResponderApplication (..),
+                                           cleanNetworkMutableState, newNetworkMutableState,
+                                           nullNetworkServerTracers, withServerNode)
+import           Ouroboros.Network.Protocol.Handshake.Codec (cborTermVersionDataCodec,
+                                                             noTimeLimitsHandshake,
+                                                             timeLimitsHandshake)
+import           Ouroboros.Network.Protocol.Handshake.Unversioned (UnversionedProtocol (..),
+                                                                   UnversionedProtocolData (..),
+                                                                   unversionedHandshakeCodec,
+                                                                   unversionedProtocolDataCodec)
+import           Ouroboros.Network.Protocol.Handshake.Type (Handshake)
+import           Ouroboros.Network.Protocol.Handshake.Version (acceptableVersion, simpleSingletonVersions)
+import           Ouroboros.Network.Util.ShowProxy (ShowProxy(..))
+
+import           Cardano.BM.Data.LogItem (LogObject)
+
+import qualified Trace.Forward.Protocol.Acceptor as Acceptor
+import qualified Trace.Forward.Protocol.Codec as Acceptor
+import           Trace.Forward.Queue (writeLogObjectsToQueue)
+import           Trace.Forward.ReqResp (Request (..), Response (..))
+import           Trace.Forward.Configuration (AcceptorConfiguration (..), HowToConnect (..))
+
+listenToForwarder
+  :: (CBOR.Serialise a,
+      ShowProxy a,
+      Typeable a)
+  => AcceptorConfiguration a
+  -> TBQueue (LogObject a)
+  -> IO Void
+listenToForwarder config loQueue = withIOManager $ \iocp -> do
+  let app = acceptorApp config loQueue
+  case forwarderEndpoint config of
+    LocalPipe localPipe -> do
+      let snocket = localSnocket iocp localPipe
+          address = localAddressFromPath localPipe
+      doListenToForwarder snocket address noTimeLimitsHandshake app
+    RemoteSocket host port -> do
+      listenAddress:_ <- Socket.getAddrInfo Nothing (Just $ T.unpack host) (Just $ show port)
+      let snocket = socketSnocket iocp
+          address = Socket.addrAddress listenAddress
+      doListenToForwarder snocket address timeLimitsHandshake app
+
+doListenToForwarder
+  :: Ord addr
+  => Snocket IO fd addr
+  -> addr
+  -> ProtocolTimeLimits (Handshake UnversionedProtocol Term)
+  -> OuroborosApplication 'ResponderMode addr LBS.ByteString IO Void ()
+  -> IO Void
+doListenToForwarder snocket address timeLimits app = do
+  networkState <- newNetworkMutableState
+  _ <- async $ cleanNetworkMutableState networkState
```

Comment:
Have you thought about async exceptions when using `async`? I don't think you need that spawned thread to run if the parent thread is killed.

Reply:
> I don't think you need that spawned thread to run if the parent thread is killed.

@coot Of course, I don't need it. Should I use spawning with automatic cancelation instead?

Reply:
You probably want `race` to avoid the situation when the `cleanNetworkMutableState` thread errored but the main thread continues.

Reply:
Thanks. I did it using `waitAnyCancel` (solution with `race_` didn't pass tests).